File: php/src/Pipeline.php
<?php declare(strict_types=1);

namespace UntangledConditionals;

use UntangledConditionals\Dependencies\Config;
use UntangledConditionals\Dependencies\Emailer;
use UntangledConditionals\Dependencies\Logger;
use UntangledConditionals\Dependencies\Project;

final class Pipeline
{
    /**
     * @var Config
     */
    private $config;

    /**
     * @var Emailer
     */
    private $emailer;

    /**
     * @var Logger
     */
    private $log;

    public function __construct(Config $config, Emailer $emailer, Logger $log)
    {
        $this->config = $config;
        $this->emailer = $emailer;
        $this->log = $log;
    }

    private function projectTestPass(Project $project) :bool {
        if (!$project->hasTests()) {
            $this->log->info('No tests');
            return true;
        }
        if (!$project->runTestsResult()) {
            $this->log->error('Tests failed');
            return false;
        }
        $this->log->info('Tests passed');
        return true;
    }

    private function projectDeployedSuccessfully(Project $project) :bool { 
        if(!$project->deploysSuccessfully()) {
            $this->log->error('Deployment failed');
            return false;
        }
        $this->log->info('Deployment successful');
        return true;
    }

    private function sendEmailNotification(string $message) :void {
        if (!$this->config->sendEmailSummary()) {
            $this->log->info('Email disabled');
            return;
        }
        $this->log->info('Sending email');
        $this->emailer->send($message);
    }

    public function run(Project $project): void
    {
        if(!$this->projectTestPass($project)) {
            $this->sendEmailNotification('Tests failed');
            return;
        }
        if(!$this->projectDeployedSuccessfully($project)) {
            $this->sendEmailNotification('Deployment failed');
            return;
        }
        $this->sendEmailNotification('Deployment completed successfully');
    }
}

File: csharp/UntangledConditionals/Pipeline.cs
using System;

namespace UntangledConditionals
{
    public class Pipeline
    {
        private readonly Config _config;
        private readonly Emailer _emailer;
        private readonly Logger _log;


        public Pipeline(Config config, Emailer emailer, Logger log)
        {
            _config = config;
            _emailer = emailer;
            _log = log;
        }
        
        public void run(Project project) {
            bool testsPassed;
            bool deploySuccessful;

            if (project.hasTests()) {
                if ("success".Equals(project.runTests())) {
                    _log.info("Tests passed");
                    testsPassed = true;
                } else {
                    _log.error("Tests failed");
                    testsPassed = false;
                }
            } else {
                _log.info("No tests");
                testsPassed = true;
            }

            if (testsPassed) {
                if ("success".Equals(project.deploy())) {
                    _log.info("Deployment successful");
                    deploySuccessful = true;
                } else {
                    _log.error("Deployment failed");
                    deploySuccessful = false;
                }
            } else {
                deploySuccessful = false;
            }

            if (_config.sendEmailSummary()) {
                _log.info("Sending email");
                if (testsPassed) {
                    if (deploySuccessful) {
                        _emailer.send("Deployment completed successfully");
                    } else {
                        _emailer.send("Deployment failed");
                    }
                } else {
                    _emailer.send("Tests failed");
                }
            } else {
                _log.info("Email disabled");
            }
        }
    }

    public interface Logger
    {
        void info(string message);
        void error(string message);
    }

    public interface Emailer
    {
        void send(string message);
    }

    public interface Config
    {
        bool sendEmailSummary();
    }
}